### DISCUSSION TITLE
Deploy an read a file on Android

### DISCUSSION THREAD

**Discussion by true-perfect-code** [dotnet/maui#11941] [Q&A] [answered]
Hi,

I want to add some settings as a file to the project and then read them when running the application. With Windows this works fine, both as an embedded resource and as content.

On Android this seems to be an impossible challenge! First, you can find a single doc for this "File system helpers" and it is very concise. Secondly, it seems that nobody asked for it during development, so you won't find any examples or answers.

What I have tried so far and what does not work:

1. I noticed that Visual Studio also has a Resources folder in the project and a Raw folder under it. I now also have the same structure under Platforms -> Android then Resources -> Raw because I don't know where the app is looking.
2. in the folder I then placed an usual text file and set build to MauiAsset
3. then I used from MAUI doc the code to read "Bundled files":
```
public async Task<string> ReadTextFile(string filePath)
{
    using Stream fileStream = await FileSystem.Current.OpenAppPackageFileAsync(filePath);
    using StreamReader reader = new StreamReader(fileStream);

    return await reader.ReadToEndAsync();
}
```
4. once I specified plain text file as filePath and when that didn't work, I also supplied the path:

`var fullpath = Path.Combine(@"Resources/Raw", @"AboutAssets.txt");`

...but that doesn't work either!

The thing that surprises me the most is that there is no error message, but the string set is equal to "" (so not even null).

Where am I making the mistake?

It must not be so difficult to include a file (whether text, Excel, image...) in the project as embedded or as content and when the application is then run on a device (whether Windows PC or Android Phone) that you then read this file and use it!?

Is there an example showing how to do this on Android (Windows is no problem)?

Thanks

**Comment by true-perfect-code** [comment, marked as answer]
Since I have lost a lot of time and there are probably still people who have or will have the same problem, here again resumed how to include a resource (text or Excel) via NET MAUI and read out on the target device (eg an Android Phone) and use in the program:

1. if one, like me, has entered an entry <ItemGroup> to the resource in cproj, delete it !
2. in the project under Project/Resources/Raw insert the required resource (I tried it only with text and Excel) and set Build Action to MauiAsset. 
3. the readout is done with the code :

```
using var stream = await FileSystem.OpenAppPackageFileAsync("FileText.txt");
using var reader = new StreamReader(stream);

var contents = reader.ReadToEnd();
Console.WriteLine(contents);
```

4. ...or if you want to read an Excel file via OpenXml, then you pass the stream to the function:
`using (SpreadsheetDocument spreadSheetDocument = SpreadsheetDocument.Open(stream, false))`

So this worked very well in the Android emulator as well as on a physical device (Android Phone). If it works on the iPhone I couldn't try it, but I would be very happy if someone could check it and write something here.

Thanks for the help.

**Comment by xtuzy** [comment]
EmbeddedResource can be set in netstandart project, file will be pack in dll, read it need correct full path, you can see ReadEmbedAssetPath method in https://github.com/xtuzy/Veldrid.Samples/blob/master/Veldrid.Maui.Samples/Veldrid.Maui.Samples.Core/LearnOpenGL/BaseGpuDrawableExtension.cs
